fields.go

```diff
@@ -33,6 +33,27 @@ type structFields struct {
 	inlinedFallback *structField
 }
 
+// reindex recomputes index to avoid bounds check during runtime.
+//
+// During the construction of each [structField] in [makeStructFields],
+// the index field is 0-indexed. However, before this returns,
+// the 0th field is stored in index0 and index stores the remainder.
```

Comment:
I assume you tried to avoid needing to do this reindex as a second step, given that it adds a bit of complexity and makes the index field have two stages with different contents.

Reply:
This approach made it easier to reason about the append logic in the meat of `makeStructFields`. I do agree though that the representation being different at different points in time can be confusing. Maybe there's a cleaner implementation, but I'll go with this for now.

Another advantage of a `reindex` method is that I didn't need to change the unit tests too much.